// anchor-link-browser-transport/lib/index.es5.js
'use strict';

var qrcode = require('qrcode');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

function __values(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
}

var styleText = "\n/* Anchor Link */\n\n.%prefix% * {\n    box-sizing: border-box;\n    line-height: 1;\n}\n\n.%prefix% {\n    font-family: -apple-system, system-ui, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue',\n        Arial, sans-serif;\n    font-size: 13px;\n    background: rgba(0, 0, 0, 0.65);\n    position: fixed;\n    top: 0px;\n    left: 0px;\n    width: 100%;\n    height: 100%;\n    z-index: 2147483647;\n    display: none;\n    align-items: center;\n    justify-content: center;\n}\n\n.%prefix%-active {\n    display: flex;\n}\n\n.%prefix%-inner {\n    background: #EFF1F7;\n    margin: 20px;\n    padding-top: 50px;\n    border-radius: 20px;\n    box-shadow: 0px 4px 100px rgba(0, 0, 0, .5);\n    width: 340px;\n    transition-property: all;\n    transition-duration: .5s;\n    transition-timing-function: ease-in-out;\n    position: relative;\n}\n\n.%prefix%-close {\n    display: block;\n    position: absolute;\n    top: 11px;\n    right: 16px;\n    width: 28px;\n    height: 28px;\n    background-image: url(\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M.57 12.1a.96.96 0 000 1.34c.37.36 1 .36 1.34 0L7 8.37l5.09 5.09c.36.35.97.35 1.34-.01a.96.96 0 000-1.34L8.34 7.01l5.08-5.08a.95.95 0 000-1.33.97.97 0 00-1.34-.01L6.99 5.68 1.91.59a.96.96 0 00-1.33 0 .97.97 0 00-.01 1.34l5.09 5.08-5.1 5.1z' fill='%23007AFF'/%3E%3C/svg%3E\");\n    background-size: 14px;\n    background-repeat: no-repeat;\n    background-position: 50% 7px;\n    border-radius: 100%;\n    cursor: pointer;\n}\n\n.%prefix%-close:hover {\n    background-color: white;\n}\n\n.%prefix%-logo {\n    width: 70px;\n    height: 70px;\n    margin: 0 auto;\n    margin-top: -56px;\n    background-image: url(\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg'%3E%3Cpath fill='%23fff' d='M18.81 9.19h33.25V59.5H18.81z'/%3E%3Cpath d='M38.45 28.88h-6.9L35 21.77l3.45 7.1z' fill='%233650A2'/%3E%3Cpath fill-rule='evenodd' clip-rule='evenodd' d='M35 70a35 35 0 100-70 35 35 0 000 70zm2.36-55.4a2.62 2.62 0 00-4.72 0L21.9 36.75h5.84l1.7-3.5h11.13l1.7 3.5h5.83L37.36 14.6zM48.13 44.2h-5.26a7.76 7.76 0 01-5.24 7v-10.5a2.62 2.62 0 10-5.25 0v10.5a7.76 7.76 0 01-5.25-7h-5.25c.16 7.06 6 12.69 13.12 12.69 7.12 0 12.97-5.63 13.13-12.7z' fill='%233650A2'/%3E%3C/svg%3E\");\n}\n\n.%prefix%-logo.loading {\n    border-radius: 100%;\n    background-color: #3650A2;\n    background-image: url(\"data:image/svg+xml,%3Csvg viewBox='0.5 0.5 45 45' xmlns='http://www.w3.org/2000/svg' stroke='%23fff'%3E%3Cg fill='none' fill-rule='evenodd' transform='translate(1 1)' stroke-width='2'%3E%3Ccircle cx='22' cy='22' r='6' stroke-opacity='0'%3E%3Canimate attributeName='r' begin='1.5s' dur='3s' values='6;22' calcMode='linear' repeatCount='indefinite' /%3E%3Canimate attributeName='stroke-opacity' begin='1.5s' dur='3s' values='1;0' calcMode='linear' repeatCount='indefinite' /%3E%3Canimate attributeName='stroke-width' begin='1.5s' dur='3s' values='2;0' calcMode='linear' repeatCount='indefinite' /%3E%3C/circle%3E%3Ccircle cx='22' cy='22' r='6' stroke-opacity='0'%3E%3Canimate attributeName='r' begin='3s' dur='3s' values='6;22' calcMode='linear' repeatCount='indefinite' /%3E%3Canimate attributeName='stroke-opacity' begin='3s' dur='3s' values='1;0' calcMode='linear' repeatCount='indefinite' /%3E%3Canimate attributeName='stroke-width' begin='3s' dur='3s' values='2;0' calcMode='linear' repeatCount='indefinite' /%3E%3C/circle%3E%3Ccircle cx='22' cy='22' r='8'%3E%3Canimate attributeName='r' begin='0s' dur='1.5s' values='6;1;2;3;4;5;6' calcMode='linear' repeatCount='indefinite' /%3E%3C/circle%3E%3C/g%3E%3C/svg%3E\");\n}\n\n.%prefix%-logo.error {\n    background-image: url(\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 70 70'%3E%3Cdefs/%3E%3Ccircle cx='35' cy='35' r='35' fill='%23FC3D39'/%3E%3Cpath fill='%23fff' d='M22.3 48h25.4c2.5 0 4-1.7 4-4a4 4 0 00-.5-2L38.5 19.3a4 4 0 00-3.5-2 4 4 0 00-3.5 2L18.8 42.1c-.3.6-.5 1.3-.5 2 0 2.2 1.6 4 4 4zM35 37c-.9 0-1.4-.6-1.4-1.5l-.2-7.7c0-.9.6-1.6 1.6-1.6s1.7.7 1.7 1.6l-.3 7.7c0 1-.5 1.5-1.4 1.5zm0 6c-1 0-1.9-.8-1.9-1.8s.9-1.8 2-1.8c1 0 1.8.7 1.8 1.8 0 1-.9 1.8-1.9 1.8z'/%3E%3C/svg%3E\");\n}\n\n.%prefix%-logo.success {\n    background-image: url(\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 70 70'%3E%3Cdefs/%3E%3Ccircle cx='35' cy='35' r='35' fill='%233DC55D'/%3E%3Cpath fill='%23fff' d='M30.9 49.7a2 2 0 001.8-1L48 24.9c.3-.5.4-1 .4-1.4 0-1-.7-1.7-1.7-1.7-.8 0-1.2.3-1.6 1L30.8 45.4 23.5 36c-.5-.6-1-.9-1.6-.9-1 0-1.8.8-1.8 1.8 0 .4.2.9.6 1.3L29 48.7c.6.7 1.1 1 1.9 1z'/%3E%3C/svg%3E\");\n}\n\n.%prefix%-request {\n    padding: 20px;\n    border-radius: 20px;\n    border-top-left-radius: 0;\n    border-top-right-radius: 0;\n    background: white;\n}\n\n.%prefix%-info {\n    display: flex;\n    flex-direction: column;\n    align-items: center;\n}\n\n.%prefix%-title {\n    color: #000000;\n    font-size: 25px;\n    margin-top: 14px;\n    font-weight: bold;\n    line-height: 30px;\n}\n\n.%prefix%-subtitle {\n    margin: 14px 0;\n    color: #5C5C5C;\n    text-align: center;\n    line-height: 1.3;\n}\n\n.%prefix%-actions {\n    display: flex;\n    flex-direction: column;\n    align-items: center;\n}\n\n.%prefix%-uri {\n    width: 100%;\n}\n\n.%prefix%-uri a {\n    color: #007AFF;\n    background: #EFF1F7;\n    text-decoration: none;\n    font-size: 17px;\n    flex-grow: 1;\n    flex: 1;\n    width: 100%;\n    line-height: 1;\n    padding: 20px 18px;\n    border-radius: 12px;\n    font-weight: 400;\n    text-align: center;\n    display: block;\n    margin-top: 1em;\n}\n\n.%prefix%-qr {\n    margin-top: 1em;\n}\n\n.%prefix%-qr svg {\n    width: 100%;\n}\n\n.%prefix%-footnote {\n    text-align: center;\n\n    width: 100%;\n    position: absolute;\n    bottom: -26px;\n    left: 0;\n    color: white;\n}\n\n.%prefix%-footnote a {\n    color: white;\n}\n\n.%prefix%-wskeepalive {\n    display: none;\n}\n\n\n\n@media (prefers-color-scheme: dark) {\n    .%prefix%-inner {\n        background: #262D43;\n        color: white;\n    }\n    .%prefix%-request {\n        background: #131B33;\n    }\n    .%prefix%-title {\n        color: #FCFCFC;\n    }\n    .%prefix%-subtitle {\n        color: #B8C0DA;\n    }\n    .%prefix%-qr svg path {\n        fill: #131B33;\n    }\n    .%prefix%-qr svg path:last-child {\n        stroke: white;\n    }\n    .%prefix%-uri a {\n        color: #FCFCFC;\n        background: #262D43;\n    }\n    .%prefix%-close {\n        background-image: url(\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M.57 12.1a.96.96 0 000 1.34c.37.36 1 .36 1.34 0L7 8.37l5.09 5.09c.36.35.97.35 1.34-.01a.96.96 0 000-1.34L8.34 7.01l5.08-5.08a.95.95 0 000-1.33.97.97 0 00-1.34-.01L6.99 5.68 1.91.59a.96.96 0 00-1.33 0 .97.97 0 00-.01 1.34l5.09 5.08-5.1 5.1z' fill='%23FCFCFC'/%3E%3C/svg%3E\");\n    }\n    .%prefix%-close:hover {\n        background-color: black;\n    }\n}\n";

var supportedChains = {
    'aca376f206b8fc25a6ed44dbdc66547c36c6c33e3a119ffbeaef943642f0e906': 'https://eos.greymass.com',
    '2a02a0053e5a8cf73a56ba0fda11e4d92e0238a4a2aa74fccf46d5a910746840': 'https://jungle3.greymass.com',
    '4667b205c6838ef70ff7988f6e8257e8be0e1284a2f59699054a018f743b1d11': 'https://telos.greymass.com',
};
function apiCall(url, body) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, fetch(url, {
                        method: 'POST',
                        headers: {
                            'Content-Type': 'application/json',
                        },
                        body: body ? JSON.stringify(body) : undefined,
                    })];
                case 1: return [2 /*return*/, (_a.sent()).json()];
            }
        });
    });
}
function fuel(request, session, updatePrepareStatus) {
    return __awaiter(this, void 0, void 0, function () {
        var cloned, chainId, nodeUrl, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    updatePrepareStatus('Detecting if Fuel is required.');
                    cloned = request.clone();
                    chainId = cloned.getChainId().toLowerCase();
                    nodeUrl = supportedChains[chainId];
                    if (!nodeUrl) {
                        throw new Error('Chain does not support Fuel.');
                    }
                    return [4 /*yield*/, apiCall(nodeUrl + '/v1/cosigner/sign', {
                            request: cloned,
                            signer: session.auth,
                        })];
                case 1:
                    result = _a.sent();
                    if (result.data.signatures[0]) {
                        cloned.setInfoKey('fuel_sig', result.data.signatures[0]);
                    }
                    else {
                        throw new Error('No signature returned from Fuel');
                    }
                    cloned.data.req = result.data.request;
                    return [2 /*return*/, cloned];
            }
        });
    });
}

var Storage = /** @class */ (function () {
    function Storage(keyPrefix) {
        this.keyPrefix = keyPrefix;
    }
    Storage.prototype.write = function (key, data) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                localStorage.setItem(this.storageKey(key), data);
                return [2 /*return*/];
            });
        });
    };
    Storage.prototype.read = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, localStorage.getItem(this.storageKey(key))];
            });
        });
    };
    Storage.prototype.remove = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                localStorage.removeItem(this.storageKey(key));
                return [2 /*return*/];
            });
        });
    };
    Storage.prototype.storageKey = function (key) {
        return this.keyPrefix + "-" + key;
    };
    return Storage;
}());
var BrowserTransport = /** @class */ (function () {
    function BrowserTransport(options) {
        if (options === void 0) { options = {}; }
        this.options = options;
        this.classPrefix = options.classPrefix || 'anchor-link';
        this.injectStyles = !(options.injectStyles === false);
        this.requestStatus = !(options.requestStatus === false);
        this.fuelEnabled = options.disableGreymassFuel !== true;
        this.storage = new Storage(options.storagePrefix || 'anchor-link');
    }
    BrowserTransport.prototype.closeModal = function () {
        this.hide();
        if (this.activeCancel) {
            this.activeRequest = undefined;
            this.activeCancel('Modal closed');
            this.activeCancel = undefined;
        }
    };
    BrowserTransport.prototype.setupElements = function () {
        var _this = this;
        if (this.injectStyles && !this.styleEl) {
            this.styleEl = document.createElement('style');
            this.styleEl.type = 'text/css';
            var css = styleText.replace(/%prefix%/g, this.classPrefix);
            this.styleEl.appendChild(document.createTextNode(css));
            document.head.appendChild(this.styleEl);
        }
        if (!this.containerEl) {
            this.containerEl = this.createEl();
            this.containerEl.className = this.classPrefix;
            this.containerEl.onclick = function (event) {
                if (event.target === _this.containerEl) {
                    event.stopPropagation();
                    _this.closeModal();
                }
            };
            document.body.appendChild(this.containerEl);
        }
        if (!this.requestEl) {
            var wrapper = this.createEl({ class: 'inner' });
            var closeButton = this.createEl({ class: 'close' });
            closeButton.onclick = function (event) {
                event.stopPropagation();
                _this.closeModal();
            };
            this.requestEl = this.createEl({ class: 'request' });
            wrapper.appendChild(this.requestEl);
            wrapper.appendChild(closeButton);
            this.containerEl.appendChild(wrapper);
        }
    };
    BrowserTransport.prototype.createEl = function (attrs) {
        var e_1, _a;
        if (!attrs)
            attrs = {};
        var el = document.createElement(attrs.tag || 'div');
        if (attrs) {
            try {
                for (var _b = __values(Object.keys(attrs)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var attr = _c.value;
                    var value = attrs[attr];
                    switch (attr) {
                        case 'src':
                            el.setAttribute(attr, value);
                            break;
                        case 'tag':
                            break;
                        case 'text':
                            el.appendChild(document.createTextNode(value));
                            break;
                        case 'class':
                            el.className = this.classPrefix + "-" + value;
                            break;
                        default:
                            el.setAttribute(attr, value);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return el;
    };
    BrowserTransport.prototype.hide = function () {
        if (this.containerEl) {
            this.containerEl.classList.remove(this.classPrefix + "-active");
        }
        this.clearTimers();
    };
    BrowserTransport.prototype.show = function () {
        if (this.containerEl) {
            this.containerEl.classList.add(this.classPrefix + "-active");
        }
    };
    BrowserTransport.prototype.displayRequest = function (request) {
        return __awaiter(this, void 0, void 0, function () {
            var sameDeviceRequest, sameDeviceUri, crossDeviceUri, isIdentity, title, subtitle, qrEl, _a, error_1, linkEl, linkA, iframe, infoEl, infoTitle, infoSubtitle, actionEl, footnoteEl, footnoteLink, footnoteLink, logoEl;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        this.setupElements();
                        sameDeviceRequest = request.clone();
                        sameDeviceRequest.setInfoKey('same_device', true);
                        sameDeviceRequest.setInfoKey('return_path', returnUrl());
                        sameDeviceUri = sameDeviceRequest.encode(true, false);
                        crossDeviceUri = request.encode(true, false);
                        isIdentity = request.isIdentity();
                        title = isIdentity ? 'Login' : 'Sign';
                        subtitle = 'Scan the QR-code with your Anchor app.';
                        qrEl = this.createEl({ class: 'qr' });
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        _a = qrEl;
                        return [4 /*yield*/, qrcode.toString(crossDeviceUri, {
                                margin: 0,
                                errorCorrectionLevel: 'L',
                            })];
                    case 2:
                        _a.innerHTML = _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _b.sent();
                        console.warn('Unable to generate QR code', error_1);
                        return [3 /*break*/, 4];
                    case 4:
                        linkEl = this.createEl({ class: 'uri' });
                        linkA = this.createEl({
                            tag: 'a',
                            href: crossDeviceUri,
                            text: 'Open Anchor app',
                        });
                        linkA.addEventListener('click', function (event) {
                            event.preventDefault();
                            if (navigator.userAgent.toLowerCase().indexOf('firefox') > -1) {
                                iframe.setAttribute('src', sameDeviceUri);
                            }
                            else {
                                window.location.href = sameDeviceUri;
                            }
                        });
                        linkEl.appendChild(linkA);
                        iframe = this.createEl({
                            class: 'wskeepalive',
                            src: 'about:blank',
                            tag: 'iframe',
                        });
                        linkEl.appendChild(iframe);
                        infoEl = this.createEl({ class: 'info' });
                        infoTitle = this.createEl({ class: 'title', tag: 'span', text: title });
                        infoSubtitle = this.createEl({ class: 'subtitle', tag: 'span', text: subtitle });
                        infoEl.appendChild(infoTitle);
                        infoEl.appendChild(infoSubtitle);
                        actionEl = this.createEl({ class: 'actions' });
                        actionEl.appendChild(qrEl);
                        actionEl.appendChild(linkEl);
                        if (isIdentity) {
                            footnoteEl = this.createEl({ class: 'footnote', text: "Don't have Anchor? " });
                            footnoteLink = this.createEl({
                                tag: 'a',
                                target: '_blank',
                                href: 'https://greymass.com/anchor',
                                text: 'Download Anchor app now',
                            });
                            footnoteEl.appendChild(footnoteLink);
                        }
                        else {
                            footnoteEl = this.createEl({
                                class: 'footnote',
                                text: 'Anchor signing is brought to you by ',
                            });
                            footnoteLink = this.createEl({
                                tag: 'a',
                                target: '_blank',
                                href: 'https://greymass.com',
                                text: 'Greymass',
                            });
                            footnoteEl.appendChild(footnoteLink);
                        }
                        emptyElement(this.requestEl);
                        logoEl = this.createEl({ class: 'logo' });
                        this.requestEl.appendChild(logoEl);
                        this.requestEl.appendChild(infoEl);
                        this.requestEl.appendChild(actionEl);
                        this.requestEl.appendChild(footnoteEl);
                        this.show();
                        return [2 /*return*/];
                }
            });
        });
    };
    BrowserTransport.prototype.showLoading = function () {
        return __awaiter(this, void 0, void 0, function () {
            var infoEl, infoTitle, infoSubtitle, logoEl;
            return __generator(this, function (_a) {
                this.setupElements();
                emptyElement(this.requestEl);
                infoEl = this.createEl({ class: 'info' });
                infoTitle = this.createEl({ class: 'title', tag: 'span', text: 'Loading' });
                infoSubtitle = this.createEl({
                    class: 'subtitle',
                    tag: 'span',
                    text: 'Preparing request...',
                });
                this.prepareStatusEl = infoSubtitle;
                infoEl.appendChild(infoTitle);
                infoEl.appendChild(infoSubtitle);
                logoEl = this.createEl({ class: 'logo loading' });
                this.requestEl.appendChild(logoEl);
                this.requestEl.appendChild(infoEl);
                this.show();
                return [2 /*return*/];
            });
        });
    };
    BrowserTransport.prototype.onRequest = function (request, cancel) {
        this.activeRequest = request;
        this.activeCancel = cancel;
        this.displayRequest(request).catch(cancel);
    };
    BrowserTransport.prototype.onSessionRequest = function (session, request, cancel) {
        if (session.metadata.sameDevice) {
            request.setInfoKey('return_path', returnUrl());
        }
        if (session.type === 'fallback') {
            this.onRequest(request, cancel);
            if (session.metadata.sameDevice) {
                // trigger directly on a fallback same-device session
                window.location.href = request.encode();
            }
            return;
        }
        this.activeRequest = request;
        this.activeCancel = cancel;
        this.setupElements();
        var timeout = session.metadata.timeout || 60 * 1000 * 2;
        var deviceName = session.metadata.name;
        var start = Date.now();
        var infoTitle = this.createEl({ class: 'title', tag: 'span', text: 'Sign' });
        var updateCountdown = function () {
            var timeLeft = timeout + start - Date.now();
            var timeFormatted = timeLeft > 0 ? new Date(timeLeft).toISOString().substr(14, 5) : '00:00';
            infoTitle.textContent = "Sign - " + timeFormatted;
        };
        this.countdownTimer = setInterval(updateCountdown, 500);
        updateCountdown();
        var infoEl = this.createEl({ class: 'info' });
        infoEl.appendChild(infoTitle);
        var subtitle;
        if (deviceName && deviceName.length > 0) {
            subtitle = "Please open Anchor app on \u201C" + deviceName + "\u201D to review and sign the transaction.";
        }
        else {
            subtitle = 'Please review and sign the transaction in the linked wallet.';
        }
        var infoSubtitle = this.createEl({ class: 'subtitle', tag: 'span', text: subtitle });
        infoEl.appendChild(infoSubtitle);
        emptyElement(this.requestEl);
        var logoEl = this.createEl({ class: 'logo' });
        this.requestEl.appendChild(logoEl);
        this.requestEl.appendChild(infoEl);
        this.show();
        if (isAppleHandheld() && session.metadata.sameDevice) {
            window.location.href = 'anchor://link';
        }
    };
    BrowserTransport.prototype.clearTimers = function () {
        if (this.closeTimer) {
            clearTimeout(this.closeTimer);
            this.closeTimer = undefined;
        }
        if (this.countdownTimer) {
            clearTimeout(this.countdownTimer);
            this.countdownTimer = undefined;
        }
    };
    BrowserTransport.prototype.updatePrepareStatus = function (message) {
        if (this.prepareStatusEl) {
            this.prepareStatusEl.textContent = message;
        }
    };
    BrowserTransport.prototype.prepare = function (request, session) {
        return __awaiter(this, void 0, void 0, function () {
            var result, timeout, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.showLoading();
                        if (!this.fuelEnabled || !session || request.isIdentity()) {
                            // don't attempt to cosign id request or if we don't have a session attached
                            return [2 /*return*/, request];
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        result = fuel(request, session, this.updatePrepareStatus.bind(this));
                        timeout = new Promise(function (r) { return setTimeout(r, 3500); }).then(function () {
                            throw new Error('Fuel API timeout after 3500ms');
                        });
                        return [4 /*yield*/, Promise.race([result, timeout])];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        error_2 = _a.sent();
                        console.info("Not applying fuel (" + error_2.message + ")");
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, request];
                }
            });
        });
    };
    BrowserTransport.prototype.onSuccess = function (request) {
        var _this = this;
        if (request === this.activeRequest) {
            this.clearTimers();
            if (this.requestStatus) {
                this.setupElements();
                var infoEl = this.createEl({ class: 'info' });
                var logoEl = this.createEl({ class: 'logo' });
                logoEl.classList.add('success');
                var infoTitle = this.createEl({ class: 'title', tag: 'span', text: 'Success!' });
                var subtitle = request.isIdentity() ? 'Identity signed.' : 'Transaction signed.';
                var infoSubtitle = this.createEl({ class: 'subtitle', tag: 'span', text: subtitle });
                infoEl.appendChild(infoTitle);
                infoEl.appendChild(infoSubtitle);
                emptyElement(this.requestEl);
                this.requestEl.appendChild(logoEl);
                this.requestEl.appendChild(infoEl);
                this.show();
                this.closeTimer = setTimeout(function () {
                    _this.hide();
                }, 1.5 * 1000);
            }
            else {
                this.hide();
            }
        }
    };
    BrowserTransport.prototype.onFailure = function (request, error) {
        if (request === this.activeRequest && error['code'] !== 'E_CANCEL') {
            this.clearTimers();
            if (this.requestStatus) {
                this.setupElements();
                var infoEl = this.createEl({ class: 'info' });
                var logoEl = this.createEl({ class: 'logo' });
                logoEl.classList.add('error');
                var infoTitle = this.createEl({
                    class: 'title',
                    tag: 'span',
                    text: 'Transaction error',
                });
                var infoSubtitle = this.createEl({
                    class: 'subtitle',
                    tag: 'span',
                    text: error.message || String(error),
                });
                infoEl.appendChild(infoTitle);
                infoEl.appendChild(infoSubtitle);
                emptyElement(this.requestEl);
                this.requestEl.appendChild(logoEl);
                this.requestEl.appendChild(infoEl);
                this.show();
            }
            else {
                this.hide();
            }
        }
    };
    return BrowserTransport;
}());
function emptyElement(el) {
    while (el.firstChild) {
        el.removeChild(el.firstChild);
    }
}
var returnUrlAlphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
var returnUrlAlphabetLen = returnUrlAlphabet.length;
/** Generate a return url with a random #fragment so mobile safari will redirect back w/o reload. */
function returnUrl() {
    var rv = window.location.href.split('#')[0] + '#';
    for (var i = 0; i < 8; i++) {
        rv += returnUrlAlphabet.charAt(Math.floor(Math.random() * returnUrlAlphabetLen));
    }
    return rv;
}
function isAppleHandheld() {
    return /iP(ad|od|hone)/i.test(navigator.userAgent);
}

module.exports = BrowserTransport;
