// ual-token-pocket/dist/TokenPocketUser.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenPocketUser = void 0;
const tp_eosjs_1 = __importDefault(require("tp-eosjs"));
const universal_authenticator_library_1 = require("universal-authenticator-library");
const UALTokenPocketError_1 = require("./UALTokenPocketError");
class TokenPocketUser extends universal_authenticator_library_1.User {
    constructor(chain, wallet) {
        super();
        this.keys = [];
        this.chainId = '';
        this.wallet = wallet;
        if (chain && chain.chainId) {
            this.chainId = chain.chainId;
        }
    }
    signTransaction(transaction, _config) {
        return __awaiter(this, void 0, void 0, function* () {
            let response;
            try {
                response = yield tp_eosjs_1.default.pushEosAction(Object.assign(Object.assign({}, transaction), { account: this.wallet.name, address: this.wallet.address }));
                if (response.result) {
                    return {
                        wasBroadcast: true,
                        transactionId: response.data.transactionId,
                        transaction,
                    };
                }
                else {
                    throw new Error('No result returned');
                }
            }
            catch (e) {
                throw new UALTokenPocketError_1.UALTokenPocketError('Unable to sign the given transaction', universal_authenticator_library_1.UALErrorType.Signing, e);
            }
        });
    }
    signArbitrary(publicKey, data, _helpText) {
        return __awaiter(this, void 0, void 0, function* () {
            let response;
            try {
                response = yield tp_eosjs_1.default.eosAuthSign({ from: this.wallet.name, publicKey, signdata: data });
                if (response.result) {
                    return response.data.signature;
                }
                else {
                    throw new Error('No result returned');
                }
            }
            catch (e) {
                throw new UALTokenPocketError_1.UALTokenPocketError('Unable to sign arbitrary string', universal_authenticator_library_1.UALErrorType.Signing, e);
            }
        });
    }
    verifyKeyOwnership(_) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error('Token Pocket does not currently support verifyKeyOwnership');
        });
    }
    getAccountName() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.wallet.name;
        });
    }
    getChainId() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.chainId;
        });
    }
    getKeys() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.keys.length === 0) {
                this.keys.push(this.wallet.address);
            }
            return this.keys;
        });
    }
}
exports.TokenPocketUser = TokenPocketUser;
