// ual-reactjs-renderer/dist/styles/authenticator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.errored = exports.chevron = exports.authTextFont = exports.authText = exports.authIconWrapper = exports.authIcon = exports.buttonHover = exports.authButton = void 0;

var _base = require("./base");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var authButton = {
  fontWeight: '100',
  color: '#ffffff',
  borderRadius: '6px',
  width: '100%',
  margin: '20px auto',
  boxShadow: '0px 0px 5px rgba(0,0,0,0.1)',
  opacity: '1',
  transition: '0.3s',
  WebkitTransition: '0.3s',
  outline: 'none'
};
exports.authButton = authButton;
var buttonHover = {
  cursor: 'pointer',
  transform: 'scale(1.03)'
};
exports.buttonHover = buttonHover;
var authIcon = {
  maxHeight: '30px',
  maxWidth: '100%',
  margin: 'auto',
  marginTop: '7%',
  display: 'block'
};
exports.authIcon = authIcon;

var authIconWrapper = _objectSpread(_objectSpread({}, _base.base), {}, {
  display: 'inline-block',
  "float": 'left',
  width: '50px',
  padding: '8.5px 10px',
  backgroundColor: 'rgba(0,0,0,0.15)',
  borderRadius: '5px 0px 0px 5px'
});

exports.authIconWrapper = authIconWrapper;

var authText = _objectSpread(_objectSpread({}, _base.base), {}, {
  display: 'inline-block',
  padding: '12px 13px 12px 15px',
  width: 'calc(100% - 50px)'
});

exports.authText = authText;
var authTextFont = {
  fontSize: '20px',
  fontWeight: 'bold',
  letterSpacing: '1.1px'
};
exports.authTextFont = authTextFont;
var chevron = {
  "float": 'right',
  height: '16px',
  marginTop: '2px'
};
exports.chevron = chevron;
var errored = {
  color: 'rgb(144, 150, 168)',
  fontSize: '1.5rem'
};
exports.errored = errored;
