// ual-reactjs-renderer/dist/styles/loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadingElementCSS = exports.loadingElementTwo = exports.loadingElementOne = exports.loadingElementThree = exports.loadingIconWithContainer = exports.loadingIcon = void 0;

var _base = require("./base");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var loadingIcon = _objectSpread(_objectSpread({}, _base.base), {}, {
  width: '41px',
  textAlign: 'center',
  marginTop: '-4px'
});

exports.loadingIcon = loadingIcon;

var loadingIconWithContainer = _objectSpread(_objectSpread({}, loadingIcon), {}, {
  width: '100%',
  padding: '20px'
});

exports.loadingIconWithContainer = loadingIconWithContainer;
var loadingElementThree = {
  width: '9px',
  height: '9px',
  margin: '1px',
  backgroundColor: 'white',
  borderRadius: '100%',
  display: 'inline-block',
  WebkitAnimation: 'ual-bouncedelay 1.4s infinite ease-in-out both',
  animation: 'ual-bouncedelay 1.4s infinite ease-in-out both'
};
exports.loadingElementThree = loadingElementThree;

var loadingElementOne = _objectSpread(_objectSpread({}, loadingElementThree), {}, {
  WebkitAnimationDelay: '-0.32s',
  AnimationDelay: '-0.32s'
});

exports.loadingElementOne = loadingElementOne;

var loadingElementTwo = _objectSpread(_objectSpread({}, loadingElementThree), {}, {
  WebkitAnimationDelay: '-0.16s',
  AnimationDelay: '-0.16s'
});

exports.loadingElementTwo = loadingElementTwo;
var loadingElementCSS = "\n  @-webkit-keyframes ual-bouncedelay {\n    0%, 80%, 100% { -webkit-transform: scale(0) }\n    40% { -webkit-transform: scale(1.0) }\n  }\n\n  @keyframes ual-bouncedelay {\n    0%, 80%, 100% { \n      -webkit-transform: scale(0);\n      transform: scale(0);\n    } 40% { \n      -webkit-transform: scale(1.0);\n      transform: scale(1.0);\n    }\n  }\n";
exports.loadingElementCSS = loadingElementCSS;
